File: motmot/polygon_map.c
// -*- coding: utf-8 -*-

#include <stdio.h>
#include <stddef.h>
#include "queue.h"


typedef struct RaggedArray
{
  void * flat;
  int itemsize;
  int length;
  int * starts;
  int * ends;
} RaggedArray;


void populate_polygon_map(ptrdiff_t * polygon_map, ptrdiff_t * faces,
                          int n, int d,
                          RaggedArray * dest_vertices, RaggedArray * polygons) {

  // The algorithm this belongs to is 2-part with the first half being written
  // in Python in ``_polygon_map.py``. See there.

  // Pull some variable getters out of the loop.
  ptrdiff_t * dest_vertices_flat = (ptrdiff_t *) dest_vertices -> flat;
  int * rag_starts = dest_vertices -> starts;
  int * rag_ends = dest_vertices -> ends;
  ptrdiff_t * polygons_flat = (ptrdiff_t *) polygons -> flat;

  // For each cell in the ``polygon_map``.
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < d; j++) {

      // Get the edge to search for.
      // We have to do some pointer arithmetic because ``faces`` is a 2D
      // array in Python but 1D in C.
      ptrdiff_t * faces_row = faces + (d * i);
      ptrdiff_t v0 = faces_row[j];
      ptrdiff_t v1 = faces_row[(j + 1) % d];

      // Find, and default to -1, where we'll write the answer to.
      ptrdiff_t * neighbour = polygon_map + (d * i) + j;
      *neighbour = -1;

      // Search for an edge which starts at v1,
      for (int k = rag_starts[v1]; k < rag_ends[v1]; k++) {
        // and ends at v0.
        if (dest_vertices_flat[k] == v0) {
          *neighbour = polygons_flat[k];
          break;
        }
      }
    }
  }
}


ptrdiff_t ravel(ptrdiff_t i, ptrdiff_t j, ptrdiff_t d) {
  /* Convert 2D (n, d) array index to ravelled 1D index. */
  return i * d + j;
}


void connected(ptrdiff_t * initial, ptrdiff_t len_initial, ptrdiff_t d,
               ptrdiff_t * polygon_map, Queue * queue) {
/*
    Navigate `polygon_map` to create a connected region.

    If this function was recursive then the recursion depths required to get
    around a typical sized mesh would cause StackOverflowErrors.
    To avoid this, recursion is mimicked using a queue.
    This `queue` is just an integer array listing `arg`s in the order they are
    encountered, with a `reverse_queue` that lists an `arg`'s position in
    `queue` to facilitate `arg in queue` checking.

    In order to avoid having to inc-ref arrays, `queue` and `reverse_queue` are
    intended to be initialised in Python using numpy.

    An arg is only put into `queue` if it is to be included in the resulting
    region so an output array is not needed as Python can just look at the queue
    to see what was used.
 */

  // Add each `arg` from `args` to the queue.
  Q_appends(queue, initial, len_initial);

  // While there are unprocessed elements in the queue.
  while (!Q_is_empty(queue)) {

    // Read the oldest unprocessed value in the queue.
    ptrdiff_t arg = Q_consume_later(queue);

    // For each adjacent polygon:
    for (ptrdiff_t i = 0; i < d; i++) {

      // `polygon_map` in Python has shape `(n, d)` but is flattened by C.
      // The index must be ravelled to match.
      ptrdiff_t index = ravel(arg, i, d);

      // Lookup the arg of an adjacent polygon.
      ptrdiff_t arg_ = polygon_map[index];

      // Not allowed or missing adjacent polygons are masked by
      // -1s in `polygon_map`. These should be skipped.
      if (arg_ == -1) continue;

      // Finally Q_add `arg_` to the queue.
      // This is ignored if `arg_` is already in it.
      Q_add(queue, arg_);
    }
  }
}


ptrdiff_t group_connected(ptrdiff_t * polygon_map, ptrdiff_t * shape,
                          ptrdiff_t * group_ids, Queue * queue) {
  /* Split a disjointed mesh into connected sub-meshes. */

  // This works simply by calling connected() repeatedly until every polygon
  // has been allocated a group number.

  ptrdiff_t group_id = 0;
  for (ptrdiff_t polygon_id = 0; polygon_id < shape[0]; polygon_id++) {

    // Skip if this polygon is already in a group.
    if (group_ids[polygon_id] != -1) continue;

    // Find all polygons connected to `polygon_id`.
    ptrdiff_t start = queue -> append_index;
    connected(&polygon_id, 1, shape[1], polygon_map, queue);
    ptrdiff_t end = queue -> append_index;

    // Unpack the unusual output of connected().
    for (ptrdiff_t i = start; i != end; i = (i + 1) % queue -> max_size) {

      ptrdiff_t _arg = queue -> queue[i];
      group_ids[_arg] = group_id;
      // Reset this queue space.
      queue -> reverse_queue[_arg] = -1;
    }

    group_id ++;
  }
  return group_id;
}

File: motmot/_polygon_map.py
# -*- coding: utf-8 -*-
"""
"""

import numpy as np
from cslug import ptr
from rockhopper import RaggedArray

from motmot._slug import slug


def make_polygon_map(faces: np.ndarray, id_count: int, use_C=True):
    """Create an array mapping each polygon to its neighbours."""
    # This pivots around the fact that if polygon p0 shares an edge v0->v1 with
    # polygon p1, then p1 contains the edge v1->v0 (hard-coding the assumption
    # that a polygon's vertices are consistently listed in anticlockwise order).
    # So to find a polygon's neighbour on a given edge, we need to find that
    # edge reversed in another polygon.

    n, d = faces.shape

    # Normally, we'd tackle this with an {(v0, v1): polygon_id} dictionary but
    # that is far too slow. Instead, group by v0 into ragged arrays.

    _polygons, _sub_ids = np.divmod(np.arange(faces.size, dtype=np.intp),
                                    faces.shape[1])
    _dest_vertices = faces[_polygons, (1 + _sub_ids) % d]

    polygons, dest_vertices = \
        RaggedArray.groups_by(faces.ravel(), _polygons, _dest_vertices,
                              id_max=id_count)

    # Now ``polygons[v0]`` is a list of polygon ids containing v0.
    # And ``dest_vertices[v0]`` is a list of all the v1s that v0 links to.

    # From here it's just a case of searching ``dest_vertices[v1]`` for v0, then
    # checking in ``polygons[v1]`` for which polygon the v0 corresponds to. This
    # step has been translated into C for speed.

    polygon_map = np.empty(faces.shape, dtype=np.intp, order="C")
    if use_C:
        populate_polygon_map_c(polygon_map, faces, n, d, dest_vertices,
                               polygons)
    else:
        populate_polygon_map_py(polygon_map, faces, n, d, dest_vertices,
                                polygons)

    return polygon_map


def populate_polygon_map_c(polygon_map, faces, n, d, dest_vertices: RaggedArray,
                           polygons: RaggedArray):
    assert faces.dtype == np.intp
    assert faces.flags.c_contiguous
    slug.dll.populate_polygon_map(ptr(polygon_map), ptr(faces), n, d,
                                  dest_vertices._c_struct._ptr,
                                  polygons._c_struct._ptr)


def populate_polygon_map_py(polygon_map, faces, n, d, dest_vertices, polygons):
    """A pure-Python proof-of-concept/test for :meth:`populate_polygon_map_c`.
    """
    for i in range(n):
        for j in range(d):
            v0 = faces[i, j]
            v1 = faces[i, (j + 1) % d]
            for k in range(dest_vertices.starts[v1], dest_vertices.ends[v1]):
                if dest_vertices.flat[k] == v0:
                    # Note that ``polygons.starts`` is ``dest_vertices.starts``
                    # and similarly for ``xxx.ends``, which is why it is ok
                    # to use indices ``k`` for ``dest_vertices`` with
                    # ``polygons`` without any kind of conversion or remapping.
                    polygon_map[i, j] = polygons.flat[k]
                    break
            else:
                polygon_map[i, j] = -1


if __name__ == "__main__":
    faces = np.array([[0, 1, 2, 5], [3, 4, 0, 5], [2, 1, 0, 6]], dtype=np.intp)
    id_count = faces.max() + 1
    target = np.array([[2, 2, -1, 1], [-1, -1, 0, -1], [0, 0, -1, -1]])
    assert np.array_equal(make_polygon_map(faces, id_count, use_C=True), target)
